File: julia/Euler/src/bin/p0021.jl

# project euler: problem 21

module Prob0021

import ..Util: get_σ_tbl

function solve_0021(limit = 10_000)
    upper = limit - 1
    d_tbl = get_σ_tbl(1, upper)
    for x = 1:upper
        d_tbl[x] -= x
    end

    sum(x + d_tbl[x] for x = 2:upper if x > d_tbl[x] && d_tbl[d_tbl[x]] == x)
end

end #module

using .Prob0021: solve_0021
export solve_0021

File: clojure/src/project_euler/sol/p0021.clj
(ns project-euler.sol.p0021
  (:require
   [project-euler.lib.math.prime :as prime]))

(defn- make-aliquot-sum-tbl
  [upper]
  (->> (prime/make-sigma-tbl 1 (dec upper))
       (map-indexed #(- %2 %1))
       (vec)))

(defn solve
  ([]
   (solve 10000))
  ([upper]
   (let [tbl (make-aliquot-sum-tbl upper)
         amicable? (fn [n] (and (> n (nth tbl n)) (= n (nth tbl (nth tbl n)))))]
     (->> (range 2 upper)
          (map #(if (amicable? %) (+ % (nth tbl %)) 0))
          (apply +)))))

